README.rst
mt103
=====

Parse MT103 messages from the Swift payments network

What's an MT103?
----------------

Banks don't really deal with cash much any more.  Instead, they push bits
around the internet tracking where your money goes digitally.  The network that
handles much of that movement is called `Swift`_, and the transfers are
documented in a special format native to that network called *MT103*.

.. _Swift: https://en.wikipedia.org/wiki/ISO_9362


What's this Do?
---------------

Unfortunately, MT103 isn't a common standard for most software developers.
It's ugly & hard to read for humans and not at all easy to parse.  This library
attempts to fix that, so all you have to do is pass an MT103 string into it and
you get back a native Python object with the properties you're looking for.

.. code-block:: python

    from mt103 import MT103

    mt103 = MT103("some-mt-103-string")
    print("basic header: {}, bank op code: {}, complete message: {}".format(
        mt103.basic_header,
        mt103.text.bank_operation_code,
        mt103
    ))


Installation
------------

It's on PyPi, so just install it with pip.

.. code-block:: shell

    $ pip install mt103


TODO
----

Parsing MT103 messages should work just fine and you should be able to access
all of the components via the Python API *except* for section ``13C``.  From
the specs I've seen, it's unclear as to whether this section is permitted to
repeat (meaning it should be parsed as a list) or if it's one value only.  If
someone can explain this authoritatively to me, I can include support for this
section as well.

CHANGELOG.md
# Changelog

## 1.1.0

* Added support for a few more portions of the `text` section:

    * instruction_code
    * transaction_type_code
    * exchange_rate
    * sending_institution
    * sender_charges
    * receiver_charges

  Big thanks to [Broto](https://github.com/bbroto06) again for reporting the
  problem.


## 1.0.1

* Bugfix [#3](https://github.com/danielquinn/mt103/issues/3): account for
application headers for outgoing messages, MUR values containing spaces, and
trailer sections containing `{}`.  Thanks to [Broto Bhattacharjee](https://github.com/bbroto06)
for the report!


## 1.0.0

* Changed the nature of the `.user_header` attribute from a string to a
  `UserHeader` object.  This new object has the same string representation
  (`str(mt103.user_header)`), but now also possesses new sub-attributes.
* Added support for user header fields including `bank_priority_code` (`bpc`),
  `message_user_reference` (`mur`), `service_type_identifier` (`sti`), and
  `unique_end_to_end_transaction_reference` (`uetr`).


## 0.0.1

Initial release.


tests.py
from datetime import date
from unittest import TestCase, mock

from mt103 import MT103, Text, UserHeader

MESSAGE1 = (
    "{1:F01ASDFJK20AXXX0987654321}"
    "{2:I103ASDFJK22XXXXN}"
    "{4: :20:20180101-ABCDEF :23B:GHIJ :32A:180117CAD5432,1 :33B:EUR9999,0 :50K:/123456-75901 SOMEWHERE New York 999999 GR :53B:/20100213012345 :57C://SC200123 :59:/201001020 First Name Last Name a12345bc6d789ef01a23 Nowhere NL :70:test reference test reason payment group: 1234567-ABCDEF :71A:SHA :77B:Test this -}"  # NOQA: E501
)
MESSAGE2 = (
    "{1:F01QWERTY22AXXX1234567890}"
    "{2:I103QWERTY33XXXXA7}"
    "{3:{108:MT103}}"
    "{4:\n:20:1234567-8901\n:23B:ABCD\n:32A:000625EUR1000,00\n:33B:EUR1000,00\n:50K:COMPANY NAME\nNAPLES\n:52A:ABCDEFGH123\n:53A:ABCDEF12\n:54A:ABCDEF1G\n:57A:ABCDEFGHIJK\n:59:/20061120050500001A01234\nBENEFICIARY NAME\n:70:/REMITTANCE INFO\n:71A:SHA\n-}"  # NOQA: E501
)
MESSAGE3 = (
    "{1:F01QWERTY22AXXX1234567890}"
    "{2:I103QWERTY33XXXXA7}"
    "{3:{113:SEPA}{111:001}{121:d2d62e74-4f7d-45dc-a230-85fa259e1694}}"
    "{4: :20:123456-ABCDEF001 :23B:GHIJ :32A:123456GBP10000,00 :33B:GBP10000,00 :50K:/This is arbitrary text :52D:/123456-78900 More arbitrary text :53B:/12345678901234 :57C://AB123456 :59:/12345678 Even more arbitrary text :70:abc - 12.34 more txt 20190115-ABCDEF :71A:SHA :72:/INS/ABCDEF01 -}"  # NOQA: E501
)
MESSAGE4 = (
    "{1:F01AAAAGRA0AXXX0057000289}"
    "{2:O1030919010321BBBBGRA0AXXX00570001710103210920N}"
    "{3:{108:MT103 003 OF 045}{121:c8b66b47-2bd9-48fe-be90-93c2096f27d2}}"
    "{4:\n:20:5387354\n:23B:CRED\n:23E:PHOB/20.527.19.60\n:32A:000526USD1101,50\n:33B:USD1121,50\n:50K:FRANZ HOLZAPFEL GMBH\nVIENNA\n:52A:BKAUATWW\n:59:723491524\nC. KLEIN\nBLOEMENGRACHT 15\nAMSTERDAM\n:71A:SHA\n:71F:USD10,\n:71F:USD10,\n:72:/INS/CHASUS33\n-}"
    "{5:{MAC:75D138E4}{CHK:DE1B0D71FA96}}"
)


class SwiftMT103TestCase(TestCase):

    def test___init__(self):

        keys = (
            "basic_header",
            "application_header",
            "user_header",
            "text",
            "trailer",
        )

        with mock.patch.object(MT103, "_populate_by_parsing") as m:
            mt103 = MT103("test")
            self.assertFalse(mt103)
            self.assertEqual(m.call_count, 1)

        for key in keys:
            self.assertIsNone(getattr(mt103, key))
            self.assertEqual(mt103.raw, "test")

    def test__populate_by_parsing_message1(self):

        mt103 = MT103(MESSAGE1)

        self.assertTrue(mt103)
        self.assertEqual(mt103.basic_header, "F01ASDFJK20AXXX0987654321")
        self.assertEqual(mt103.application_header, "I103ASDFJK22XXXXN")
        self.assertFalse(mt103.user_header)
        self.assertEqual(mt103.trailer, None)

        self.assertEqual(str(mt103.text), MESSAGE1[54:-3])

    def test__populate_by_parsing_message2(self):

        mt103 = MT103(MESSAGE2)
        self.assertTrue(mt103)
        self.assertEqual(mt103.basic_header, "F01QWERTY22AXXX1234567890")
        self.assertEqual(mt103.application_header, "I103QWERTY33XXXXA7")
        self.assertEqual(str(mt103.user_header), "{108:MT103}")
        self.assertEqual(mt103.text.raw, MESSAGE2[70:-3])
        self.assertEqual(mt103.trailer, None)

    def test__populate_by_parsing_message3(self):

        mt103 = MT103(MESSAGE3)
        self.assertTrue(mt103)
        self.assertEqual(mt103.basic_header, "F01QWERTY22AXXX1234567890")
        self.assertEqual(mt103.application_header, "I103QWERTY33XXXXA7")
        self.assertEqual(
            str(mt103.user_header),
            "{113:SEPA}{111:001}{121:d2d62e74-4f7d-45dc-a230-85fa259e1694}"
        )
        self.assertEqual(mt103.text.raw, MESSAGE3[120:-3])
        self.assertEqual(mt103.trailer, None)

    def test__populate_by_parsing_message4(self):

        mt103 = MT103(MESSAGE4)
        self.assertTrue(mt103)
        self.assertEqual(mt103.basic_header, "F01AAAAGRA0AXXX0057000289")
        self.assertEqual(mt103.application_header, "O1030919010321BBBBGRA0AXXX00570001710103210920N")
        self.assertEqual(
            str(mt103.user_header),
            "{108:MT103 003 OF 045}{121:c8b66b47-2bd9-48fe-be90-93c2096f27d2}"
        )
        self.assertEqual(mt103.text.raw, MESSAGE4[152:-39])
        self.assertEqual(mt103.trailer, "{MAC:75D138E4}{CHK:DE1B0D71FA96}")

    def test_truthyness(self):
        self.assertFalse(MT103(""))
        self.assertFalse(MT103("test"))


class UserHeaderTestCase(TestCase):

    def test___init__(self):

        keys = (
            "bank_priority_code",
            "message_user_reference",
            "service_type_identifier",
            "unique_end_to_end_transaction_reference",
        )

        with mock.patch.object(Text, "_populate_by_parsing"):
            user_header = UserHeader("test")

        self.assertFalse(user_header)

        for key in keys:
            self.assertIsNone(getattr(user_header, key))
            self.assertEqual(user_header.raw, "test")

    def test__populate_by_parsing(self):

        mt103 = MT103(MESSAGE1)
        self.assertTrue(mt103)
        self.assertFalse(mt103.user_header)

        mt103 = MT103(MESSAGE2)
        self.assertTrue(mt103)
        self.assertTrue(mt103.user_header)
        self.assertIsNone(mt103.user_header.bank_priority_code)
        self.assertIsNone(mt103.user_header.bpc)
        self.assertEqual(mt103.user_header.message_user_reference, "MT103")
        self.assertEqual(mt103.user_header.mur, "MT103")
        self.assertIsNone(mt103.user_header.service_type_identifier)
        self.assertIsNone(mt103.user_header.sti)
        self.assertIsNone(mt103.user_header.unique_end_to_end_transaction_reference)  # NOQA: E501
        self.assertIsNone(mt103.user_header.uetr)

        mt103 = MT103(MESSAGE3)
        self.assertTrue(mt103)
        self.assertTrue(mt103.user_header)
        self.assertEqual(mt103.user_header.bank_priority_code, "SEPA")
        self.assertEqual(mt103.user_header.bpc, "SEPA")
        self.assertIsNone(mt103.user_header.message_user_reference)
        self.assertIsNone(mt103.user_header.mur)
        self.assertEqual(mt103.user_header.service_type_identifier, "001")
        self.assertEqual(mt103.user_header.sti, "001")
        self.assertEqual(
            "d2d62e74-4f7d-45dc-a230-85fa259e1694",
            mt103.user_header.unique_end_to_end_transaction_reference
        )
        self.assertEqual(
            "d2d62e74-4f7d-45dc-a230-85fa259e1694",
            mt103.user_header.uetr
        )


class TextTestCase(TestCase):

    def test___init__(self):

        keys = (
            "transaction_reference",
            "bank_operation_code",
            "interbank_settled_currency",
            "interbank_settled_amount",
            "original_ordered_currency",
            "original_ordered_amount",
            "ordering_customer",
            "ordering_institution",
            "sender_correspondent",
            "receiver_correspondent",
            "intermediary",
            "account_with_institution",
            "beneficiary",
            "remittance_information",
            "details_of_charges",
            "sender_to_receiver_information",
            "regulatory_reporting",
            "date",
        )

        with mock.patch.object(Text, "_populate_by_parsing"):
            text = Text("test")

        self.assertFalse(text)

        for key in keys:
            self.assertIsNone(getattr(text, key))
            self.assertEqual(text.raw, "test")

    def test__populate_by_parsing_message1(self):

        mt103 = MT103(MESSAGE1)

        self.assertTrue(mt103)
        self.assertTrue(mt103.text)
        self.assertEqual(mt103.text.transaction_reference, "20180101-ABCDEF")
        self.assertEqual(mt103.text.interbank_settled_currency, "CAD")
        self.assertEqual(mt103.text.original_ordered_currency, "EUR")
        self.assertEqual(mt103.text.bank_operation_code, "GHIJ")
        self.assertEqual(mt103.text.date, date(2018, 1, 17))
        self.assertEqual(
            mt103.text.ordering_customer,
            "/123456-75901 SOMEWHERE New York 999999 GR"
        )
        self.assertEqual(mt103.text.regulatory_reporting, "Test this")
        self.assertEqual(mt103.text.sender_to_receiver_information, None)
        self.assertEqual(mt103.text.original_ordered_amount, "9999,0")
        self.assertEqual(
            mt103.text.beneficiary,
            "/201001020 First Name Last Name a12345bc6d789ef01a23 Nowhere NL"
        )
        self.assertEqual(
            mt103.text.remittance_information,
            "test reference test reason payment group: 1234567-ABCDEF"
        )
        self.assertEqual(mt103.text.details_of_charges, "SHA")
        self.assertEqual(mt103.text.sender_correspondent, "/20100213012345")
        self.assertEqual(mt103.text.intermediary, None)
        self.assertEqual(mt103.text.receiver_correspondent, None)
        self.assertEqual(mt103.text.interbank_settled_amount, "5432,1")
        self.assertEqual(mt103.text.ordering_institution, None)
        self.assertEqual(mt103.text.account_with_institution, "//SC200123")

    def test__populate_by_parsing_message4(self):

        mt103 = MT103(MESSAGE4)

        self.assertTrue(mt103)
        self.assertTrue(mt103.text)
        self.assertEqual(mt103.text.transaction_reference, "5387354")
        self.assertEqual(mt103.text.interbank_settled_currency, "USD")
        self.assertEqual(mt103.text.original_ordered_currency, "USD")
        self.assertEqual(mt103.text.bank_operation_code, "CRED")
        self.assertEqual(mt103.text.date, date(2000, 5, 26))
        self.assertEqual(
            mt103.text.ordering_customer,
            "FRANZ HOLZAPFEL GMBH\nVIENNA"
        )
        self.assertIsNone(mt103.text.regulatory_reporting)
        self.assertEqual(
            mt103.text.sender_to_receiver_information,
            "/INS/CHASUS33"
        )
        self.assertEqual(mt103.text.original_ordered_amount, "1121,50")
        self.assertEqual(
            mt103.text.beneficiary,
            "723491524\nC. KLEIN\nBLOEMENGRACHT 15\nAMSTERDAM"
        )
        self.assertIsNone(mt103.text.remittance_information)
        self.assertEqual(mt103.text.details_of_charges, "SHA")
        self.assertIsNone(mt103.text.sender_correspondent)
        self.assertIsNone(mt103.text.intermediary)
        self.assertEqual(mt103.text.receiver_correspondent, None)
        self.assertEqual(mt103.text.interbank_settled_amount, "1101,50")
        self.assertEqual(mt103.text.ordering_institution, "BKAUATWW")
        self.assertIsNone(mt103.text.account_with_institution)

    def test_truthyness(self):
        self.assertFalse(Text(""))
        self.assertFalse(Text("test"))
        self.assertFalse(MT103("").text)
        self.assertFalse(MT103("test").text)
        self.assertTrue(MT103(MESSAGE1).text)


mt103.py
import re
from datetime import date


class MT103:
    """
    Parses an MT103 standard banking format string into a string-like Python
    object so you can do things like `mt103.basic_header` or `print(mt103)`.

    Usage:

    mt103 = MT103("some-mt-103-string")
    print("basic header: {}, bank op code: {}, complete message: {}".format(
        mt103.basic_header,
        mt103.text.bank_operation_code
        mt103
    ))

    With considerable help from:
    http://www.sepaforcorporates.com/swift-for-corporates/read-swift-message-structure/
    https://www.sepaforcorporates.com/swift-for-corporates/explained-swift-gpi-uetr-unique-end-to-end-transaction-reference/
    """

    MESSAGE_REGEX = re.compile(
        r"^"
        r"({1:(?P<basic_header>[^}]+)})?"
        r"({2:(?P<application_header>(I|O)[^}]+)})?"
        r"({3:"
            r"(?P<user_header>"
                r"({113:[A-Z]{4}})?"
                r"({108:[A-Z 0-9]{0,16}})?"
                r"({111:[0-9]{3}})?"
                r"({121:[a-zA-Z0-9]{8}-[a-zA-Z0-9]{4}-4[a-zA-Z0-9]{3}-[89ab][a-zA-Z0-9]{3}-[a-zA-Z0-9]{12}})?"  # NOQA: E501
            r")"
        r"})?"
        r"({4:\s*(?P<text>.+?)\s*-})?"
        r"({5:(?P<trailer>.+)})?"
        r"$",
        re.DOTALL
    )

    def __init__(self, message):

        if message is None:
            message = ""

        self.raw = message.strip()
        self.basic_header = None
        self.application_header = None
        self.user_header = None
        self.text = None
        self.trailer = None

        self._boolean = False

        self._populate_by_parsing()

    def __str__(self):
        return self.raw

    def __repr__(self):
        return str(self)

    def __bool__(self):
        return self._boolean
    __nonzero__ = __bool__  # Python 2

    def _populate_by_parsing(self):

        if not self.raw:
            return

        m = self.MESSAGE_REGEX.match(self.raw)

        self._boolean = bool(m)

        if not m:
            return None

        self.basic_header = m.group("basic_header")
        self.application_header = m.group("application_header")
        self.trailer = m.group("trailer")

        self.user_header = UserHeader(m.group("user_header"))
        self.text = Text(m.group("text") or "")


class UserHeader:
    """
    The user header is sufficiently complicated that we might want to break it
    up a bit too.
    """

    REGEX = re.compile(
        r"^"
        r"({113:(?P<bpc>[A-Z]{4})})?"
        r"({108:(?P<mur>[A-Z0-9]{0,16})})?"
        r"({111:(?P<sti>[0-9]{3})})?"
        r"({121:(?P<uetr>[a-zA-Z0-9]{8}-[a-zA-Z0-9]{4}-4[a-zA-Z0-9]{3}-[89ab][a-zA-Z0-9]{3}-[a-zA-Z0-9]{12})})?"  # NOQA: E501
        r"$"
    )

    def __init__(self, raw):

        self.raw = raw

        self.bank_priority_code = None
        self.message_user_reference = None
        self.service_type_identifier = None
        self.unique_end_to_end_transaction_reference = None

        self._boolean = False

        self._populate_by_parsing()

    def __str__(self):
        return self.raw

    def __repr__(self):
        return str(self)

    def __bool__(self):
        return self._boolean
    __nonzero__ = __bool__  # Python 2

    @property
    def bpc(self):
        return self.bank_priority_code

    @property
    def mur(self):
        return self.message_user_reference

    @property
    def sti(self):
        return self.service_type_identifier

    @property
    def uetr(self):
        return self.unique_end_to_end_transaction_reference

    def _populate_by_parsing(self):

        if not self.raw:
            return

        m = self.REGEX.match(self.raw)

        self._boolean = bool(m)

        if not m:
            return

        self.bank_priority_code = m.group("bpc")
        self.message_user_reference = m.group("mur")
        self.service_type_identifier = m.group("sti")
        self.unique_end_to_end_transaction_reference = m.group("uetr")


class Text:
    """
    With considerable help from:
      https://gist.github.com/dmcruz/9940a6b217ff701b8f3e
    """

    REGEX = re.compile(
        r"^"
        r"(:20:(?P<transaction_reference>[^\s:]+)\s*)?"
        r"(:13C:/"
          r"(?P<time_indication_class>(CLSTIME|RNCTIME|SNDTIME))/"
          r"(?P<time_indication_time>[\d]{4})"
          r"(?P<time_indication_sign>[+-])"
          r"(?P<time_indication_offset>\d{4})[^:]*"
        r")*?"
        r"(:23B:(?P<bank_operation_code>[^\s:]+)\s*)?"
        r"(:23E:(?P<instruction_code>[^:]*))?"
        r"(:26T:(?P<transaction_type_code>[^:]*))?"
        r"(:32A:"
            r"(?P<date_year>\d\d)"
            r"(?P<date_month>\d\d)"
            r"(?P<date_day>\d\d)"
            r"(?P<interbank_settled_currency>[A-Z]{3})"
            r"(?P<interbank_settled_amount>[\d,]+)"
        r"\s*)?"
        r"(:33B:"
            r"(?P<original_ordered_currency>[A-Z]{3})"
            r"(?P<original_ordered_amount>[\d,]+)"
        r"\s*)?"
        r"(:36:(?P<exchange_rate>[^:]*))?"
        r"(:50[AFK]:(?P<ordering_customer>.*?)\s*(?=(:\d\d)?))?"
        r"(:51A:(?P<sending_institution>[^:]*))?"
        r"(:52[AD]:(?P<ordering_institution>.*?)\s*(?=(:\d\d)?))?"
        r"(:53[ABD]:(?P<sender_correspondent>[^\s:]*)\s*)?"
        r"(:54[ABD]:(?P<receiver_correspondent>.*?)\s*(?=(:\d\d)?))?"
        r"(:56[ACD]:(?P<intermediary>.*?)\s*(?=(:\d\d)?))?"
        r"(:57[ABCD]:(?P<account_with_institution>.*?)\s*(?=(:\d\d)?))?"
        r"(:59A?:(?P<beneficiary>.*?)\s*(?=(:\d\d)?))?"
        r"(:70:(?P<remittance_information>.*?)\s*(?=(:\d\d)?))?"
        r"(:71A:(?P<details_of_charges>.*?)\s*(?=(:\d\d)?))?"
        r"(:71F:(?P<sender_charges>[^:]*))*"
        r"(:71G:(?P<receiver_charges>[^:]*))?"
        r"(:72:(?P<sender_to_receiver_information>.*?)\s*(?=(:\d\d)?))?"
        r"(:77B:(?P<regulatory_reporting>.*?)\s*(?=(:\d\d)?))?"
        r"$",
        re.DOTALL
    )  # NOQA

    def __init__(self, raw):

        self.raw = raw

        self.transaction_reference = None
        self.bank_operation_code = None
        self.instruction_code = None
        self.transaction_type_code = None
        self.interbank_settled_currency = None
        self.interbank_settled_amount = None
        self.original_ordered_currency = None
        self.original_ordered_amount = None
        self.exchange_rate = None
        self.ordering_customer = None
        self.sending_institution = None
        self.ordering_institution = None
        self.sender_correspondent = None
        self.receiver_correspondent = None
        self.intermediary = None
        self.account_with_institution = None
        self.beneficiary = None
        self.remittance_information = None
        self.details_of_charges = None
        self.sender_charges = None
        self.receiver_charges = None
        self.sender_to_receiver_information = None
        self.regulatory_reporting = None

        self.date = None

        self._boolean = False

        self._populate_by_parsing()

    def __str__(self):
        return self.raw

    def __repr__(self):
        return str(self)

    def __bool__(self):
        return self._boolean
    __nonzero__ = __bool__  # Python 2

    def _populate_by_parsing(self):

        if not self.raw:
            return

        m = self.REGEX.match(self.raw)

        self._boolean = bool(m)

        if not m:
            return

        for k, v in m.groupdict().items():
            if v is None:
                continue
            if k.startswith("date_"):
                continue
            if not hasattr(self, k):
                continue
            setattr(self, k, v)

        try:
            self.date = date(
                2000 + int(m.group("date_year")),
                int(m.group("date_month")),
                int(m.group("date_day"))
            )
        except (ValueError, TypeError):
            pass  # Defaults to None
